### ISSUE TITLE
Adding LimitNOFILE to systemd service file

### ISSUE THREAD

**Issue by njh** [bbc/speculate#24]
Hello,

I would like to add ```LimitNOFILE``` option to my systemd .service file to increase the number of files (sockets) that node.js can have open simultaneously:

https://www.freedesktop.org/software/systemd/man/systemd.exec.html

I am happy to submit a Pull Request if you think it makes sense - better than hacking the file post-generation I think.

Does this format look reasonable?

```json
{
  "spec": {
    "service-options": {
      "LimitNOFILE": 10000
    }
  }
}
```

nick.

**Comment by DaMouse404** [comment]
Rather than hacking the file post generation, it's probably better to use the `$UNIT.service.d` directory for these overrides as they're more about systemd than they are about node.js.

For example you can create a file named `/etc/systemd/system/$UNIT.service.d/sockets.conf`:

```
[Service]
LimitNOFILE=10000
```

**Comment by njh** [comment]
I think having multiple seperate files for a service would introduce potential for confusion - I would not expect to have to look in a seperate file for the service, containing a single option line - I would expect it all to be in the main .service file - particularly as they are not very long anyway.

Maybe I will risk writing a PR and see what you think.

**Comment by njh** [comment]
Pull request submitted as #25

**Comment by DaMouse404** [comment]
I can understand wanting to add all the systemd options into here but I'm not entirely sure it's the right place. Speculate is designed to create a basic RPM spec not to replace building up systemd units, which do use the previously mentioned pattern.

I'm also not entirely sure your approach is a good idea as it means you're going to get errors around sockets not being within Node.js rather than properly handling the connection pool?

As a final nitpick, the `service-options` parameter should really be `serviceOptions` for consistency.

**Comment by njh** [comment]
I really like that speculate is very minimal and focussed on RPM building - it is a shame that it has to be tied to systemd at all - and can't use use a generic description of the service, that something else can interpret.

This doesn't feel like about building up/layering options - it is a core part of making the service run correctly. Having a seperate file makes sense in the case that the system administrator wants to override the package provided defaults, or customise a service provided by a seperate vendor.

The other thing that made me think that this was an appropriate way to do it is the existing support for environment variables - which could equally be done using a overrides file.


I am not sure what you mean by this:
> you're going to get errors around sockets not being within Node.js rather than properly handling the connection pool?

I am trying to increase the number of available file descriptions/sockets available to the node.js process (only). You might be confused because I am working with the STOMP protocol, not HTTP.

I agree that ```service-options``` should be ```serviceOptions```.


But if you really still disagree, I will look into doing this using an overrides file - it might be worth adding instructions on doing this to the README.

**Comment by DaMouse404** [comment]
I think given the amount of change required, this is ok to go in. I've merged #26 and published as `1.7.0`. 

My other comment was that when you go to open your socket over the limit, node will return an error because it can't open any more. We encountered this and solved it by ensuring only X connections are allowed open at any given time in a queue, this could be paired with the above option to increase the size of the queue and this is potentially what you're trying to achieve?

**Comment by njh** [comment]
Thank you for accepting the PR!

Yes, I need to think about the size of the queue. When a server is rebooted/replaced, then there will be a massive rush of incoming connections from the radios - so having a good sized queue is probably a good thing.

However when I reached the File Descriptor limit, it meant that I couldn't access the HTTP admin interface - so I might need to get node.js to enforce a lower connection soft-limit before it hits the OS hard-limit.

BTW, this is my project:
https://github.com/bbc/node-radiovis-stomp-server
